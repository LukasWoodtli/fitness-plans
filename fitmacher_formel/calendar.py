from datetime import datetime

from icalendar import Event, Calendar


class CalendarEvent:
    def __init__(self, summary, date, description=""):
        self.summary = summary
        self.date = date
        self.description = description

    def get_event(self):
        event = Event()
        event.add('summary', self.summary)
        event.add('dtstart', self.date)
        event.add('dtend', self.date)
        event.add('dtstamp', datetime.now())
        event.add('description', self.description)
        return event

    def add_to_cal(self, cal):
        cal.add_component(self.get_event())


def add_to_calender(workout_events):
    cal = Calendar()

    cal.add('prodid', "Fitness")
    cal.add('version', '2.0')
    [event.add_to_cal(cal) for event in workout_events]

    return cal