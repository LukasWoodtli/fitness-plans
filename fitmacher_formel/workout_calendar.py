from datetime import datetime

from icalendar import Event
from icalendar import Calendar


class CalendarEvent:
    def __init__(self, summary, date, description=""):
        self.summary = summary
        self.date = date
        self.description = description

    def get_event_as_ical(self):
        event = Event()
        event.add('summary', self.summary)
        event.add('dtstart', self.date)
        event.add('dtend', self.date)
        event.add('dtstamp', datetime.now())
        event.add('description', self.description)
        return event


class WorkoutCalendar:
    def __init__(self, workouts):
        super().__init__()
        self.cal = Calendar()
        self._add_to_calender(workouts)

    def _add_to_calender(self, workout_events):
        self.cal.add('prodid', "Fitness")
        self.cal.add('version', '2.0')
        for event in workout_events:
            self.cal.add_component(event.get_event_as_ical())

    def get_all_as_ical(self):
        return self.cal.to_ical().decode('utf-8')