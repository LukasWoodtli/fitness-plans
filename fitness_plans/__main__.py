from fitmacher_formel import FitMacherFormel

if __name__ == "__main__":
    ff = FitMacherFormel()
    ff.create_workout_calendar()
