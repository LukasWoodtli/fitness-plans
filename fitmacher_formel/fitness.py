#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from datetime import date, timedelta
import re
import os

from fitmacher_formel.workout_calendar import CalendarEvent, Calendar

SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
INPUT_FILE = os.path.join(SCRIPT_DIR, '../dfmf.txt')

WORKOUT_START = date.today() + timedelta(days=1) #date(year=2019, month=10, day=1)
FIRST_WORKOUT = 1

BEGIN_TEXT = "GAT .1"  # just a bit obfuscated
BEGIN_TEXT = BEGIN_TEXT[::-1]

END_TEXT = "RIM TIM MUART NENIEM EBEL"  # just a bit obfuscated
END_TEXT = END_TEXT[::-1]

SPLIT_DAY_RE = re.compile(r"([0-9]{1,2}\. TAG[^\n]*)", flags=re.M)


def read_input_file():
    with open(INPUT_FILE, 'r', encoding='utf-8') as infile:
        data = infile.read()
        data = data[data.find(BEGIN_TEXT):]  # remove begin
        data = data.split(END_TEXT)[0]  # remove end
        return data


def split_workouts(data):
    workout_date = WORKOUT_START
    all_workouts = []
    sections = re.split(SPLIT_DAY_RE, data)
    sections = sections[1:]  # first entry is empty
    assert len(sections) % 2 == 0
    for s in range(0, int(len(sections)/2)):
        title = sections[s*2]
        description = title + sections[s * 2 + 1]
        description = re.sub("\n\n+", "\n\n", description)
        event = CalendarEvent(title, workout_date, description)
        all_workouts.append(event)
        workout_date = workout_date + timedelta(days=1)
    return all_workouts



def main():
    fout = open('fitness.ics', 'w')
    workouts_as_ical = create_workuts_from_text()
    fout.writelines(workouts_as_ical)
    fout.close()


def create_workuts_from_text():
    workouts = split_workouts(read_input_file())
    workouts = [workout for workout in workouts if int(workout.summary.split(".")[0]) >= FIRST_WORKOUT]
    cal = Calendar(workouts)
    workouts_as_ical = cal.get_all_as_ical()
    return workouts_as_ical


if __name__ == "__main__":
    main()
