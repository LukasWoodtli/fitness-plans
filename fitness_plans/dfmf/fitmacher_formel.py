#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import base64
import pathlib
from datetime import date, timedelta
import re
import os

from fitness_plans.lib.abstract_fitness_plan import FitnessPlan
from fitness_plans.lib.workout_calendar import CalendarEvent, WorkoutCalendar

SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))


class FitMacherFormel(FitnessPlan):
    INPUT_FILE = os.path.join(SCRIPT_DIR, 'dfmf.encoded.txt')
    CALENDAR_NAME = "DieFitMacherFormel"

    def __init__(self, workout_start_date=date.today(), first_workout=1, output_dir=None):
        super().__init__()
        self.workout_start = workout_start_date
        self.first_workout = first_workout
        self.output_dir = output_dir

    def read_input_file(self):
        with open(self.INPUT_FILE, 'rb') as infile:
            encoded_data = infile.read()
            data = base64.decodebytes(encoded_data)
            data = data.decode('UTF-8')
            return data

    def create_workouts_from_input(self, input):
        workouts = self._split_workouts(self.read_input_file())
        return workouts

    def filter_workouts(self, workouts):
        return [workout for workout in workouts if int(workout.summary.split(".")[0]) >= self.first_workout]

    def create_workout_calendar_ics(self, workouts):
        cal = WorkoutCalendar(workouts)
        workouts_as_ical = cal.get_all_as_ical()
        return workouts_as_ical

    def save_calendar(self, ics):
        output_file_path = f'{self.CALENDAR_NAME}.ics'
        if self.output_dir:
            pathlib.Path(self.output_dir).mkdir(parents=True, exist_ok=True)
            output_file_path = os.path.join(self.output_dir, output_file_path)
        with open(output_file_path, 'w') as f_out:
            f_out.writelines(ics)

    def _split_workouts(self, data):
        split_day_re = r"([0-9]{1,2}\. TAG[^\n]*)"

        all_workouts = []
        sections = re.split(split_day_re, data, flags=re.M)
        sections = sections[1:]  # first entry is empty
        assert len(sections) % 2 == 0
        for s in range(0, int(len(sections)/2)):
            event = self._create_workout_event(s, sections)
            all_workouts.append(event)
        return all_workouts

    def _create_workout_event(self, sections_index, sections):
        title_index = sections_index * 2
        title = sections[title_index]
        description = sections[title_index + 1]
        description = self._create_description(description, title)
        workout_date = self._calculate_workout_date(sections_index)
        event = CalendarEvent(title, workout_date, description)
        return event

    @staticmethod
    def _create_description(description, title):
        description = re.sub("\n\n+", "\n\n", description)
        description = title + description
        return description

    def _calculate_workout_date(self, sections_index):
        return self.workout_start + (timedelta(days=1) * sections_index)


if __name__ == "__main__":
    fitmacher_formel = FitMacherFormel()
    fitmacher_formel.build()
