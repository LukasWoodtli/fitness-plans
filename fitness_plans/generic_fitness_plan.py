from datetime import timedelta, date

from fitness_plans.lib.workout_calendar import CalendarEvent

MON = 1
TUE = 2
WED = 3
THU = 4
FRI = 5
SAT = 6
SUN = 7

workouts = {
    MON: ("Jogging", "Intervall 20 min"),
    TUE: ("Liegestütze", "10 Wiederholungen"),
    WED: ("Rumpfbeugen", "50 Wiederholungen"),
    THU: ("Klimmzüge", "5 Wiederholungen"),
    FRI: ("Yoga", ""),
    SAT: ("Joggen", "Lang"),
    SUN: ("Joggen", "Lang"),
}


class GenericWorkoutPlan:

    def __init__(self, start_date, end_date):
        self.start_date = start_date
        self.end_date = end_date

    def _create_workouts(self):
        current_date = self.start_date
        while current_date < self.end_date:
            week_day = current_date.isoweekday()
            current_workout = workouts[week_day]
            event = CalendarEvent(current_workout[0],
                          current_date,
                          current_workout[1])

            print(event.date)
            print(event.description)
            print(event.summary)
            current_date = current_date + timedelta(days=1)



if __name__ == "__main__":
    plan = GenericWorkoutPlan(date.today(), date.today() + timedelta(days=10))
    plan._create_workouts()