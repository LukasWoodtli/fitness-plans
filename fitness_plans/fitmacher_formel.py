#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from datetime import date, timedelta
import re
import os

from fitness_plans.workout_calendar import CalendarEvent, WorkoutCalendar

SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))


class FitMacherFormel:
    INPUT_FILE = os.path.join(SCRIPT_DIR, '../dfmf.txt')
    CALENDAR_NAME = "DieFitMacherFormel"

    def __init__(self, workout_start_date=date.today(), first_workout=1):
        self.workout_start = workout_start_date
        self.first_workout = first_workout

    def create_workout_calendar(self):
        with open(f'{self.CALENDAR_NAME}.ics', 'w') as f_out:
            workouts_as_ical = self._create_workouts_from_text()
            f_out.writelines(workouts_as_ical)

    def _create_workouts_from_text(self):
        workouts = self._split_workouts(self._read_input_file())
        workouts = self._filter_skipped_workouts(workouts)
        cal = WorkoutCalendar(workouts)
        return cal.get_all_as_ical()

    def _read_input_file(self):
        with open(self.INPUT_FILE, 'r', encoding='utf-8') as infile:
            data = infile.read()
            data = self._strip_unneeded_text(data)
            return data

    @staticmethod
    def _strip_unneeded_text(data):
        data = FitMacherFormel._strip_unneeded_begin_of_text(data)
        data = FitMacherFormel._strip_unneeded_end_of_text(data)
        return data

    @staticmethod
    def _strip_unneeded_end_of_text(data):
        end_text = "RIM TIM MUART NENIEM EBEL"  # just a bit obfuscated
        end_text = end_text[::-1]
        data = data.split(end_text)[0]  # remove end
        return data

    @staticmethod
    def _strip_unneeded_begin_of_text(data):
        begin_text = "GAT .1"  # just a bit obfuscated
        begin_text = begin_text[::-1]
        data = data[data.find(begin_text):]  # remove begin
        return data

    def _filter_skipped_workouts(self, workouts):
        return [workout for workout in workouts if int(workout.summary.split(".")[0]) >= self.first_workout]

    def _split_workouts(self, data):
        split_day_re = r"([0-9]{1,2}\. TAG[^\n]*)"

        all_workouts = []
        sections = re.split(split_day_re, data, flags=re.M)
        sections = sections[1:]  # first entry is empty
        assert len(sections) % 2 == 0
        for s in range(0, int(len(sections)/2)):
            event = self._create_workout_event(s, sections)
            all_workouts.append(event)
        return all_workouts

    def _create_workout_event(self, sections_index, sections):
        title_index = sections_index * 2
        title = sections[title_index]
        description = sections[title_index + 1]
        description = self._create_description(description, title)
        workout_date = self._calculate_workout_date(sections_index)
        event = CalendarEvent(title, workout_date, description)
        return event

    @staticmethod
    def _create_description(description, title):
        description = re.sub("\n\n+", "\n\n", description)
        description = title + description
        return description

    def _calculate_workout_date(self, sections_index):
        return self.workout_start + (timedelta(days=1) * sections_index)


if __name__ == "__main__":
    fitmacher_formel = FitMacherFormel()
    fitmacher_formel.create_workout_calendar()
