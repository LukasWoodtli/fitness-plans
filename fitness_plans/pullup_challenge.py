#!/usr/bin/env python3
from datetime import timedelta, date, datetime
import requests
from bs4 import BeautifulSoup

from fitness_plans.workout_calendar import CalendarEvent, WorkoutCalendar


class Workout:
    def __init__(self, week, title, subtitle):
        self.week = week
        self.title = title
        self.subtitle = subtitle
        self.easy_set = []
        self.medium_set = []
        self.hard_set = []

    def get_full_title(self):
        return f"{self.title} (Week {self.week})"

    def _get_workout_set_text(self, level):
        text = f"{level.title()}:\n"

        if level.lower() == "easy":
            workout_set = self.easy_set
        elif level.lower() == "medium":
            workout_set = self.medium_set
        elif level.lower() == "hard":
            workout_set = self.hard_set
        else:
            raise Exception(f"Level not known: {level}")

        prerequisites = workout_set[0]
        text += f"Prerequisites: {prerequisites}\n\n"

        for i, s in enumerate(workout_set[1:], 1):
            text += f"\tSet {i}: {s}\n"
        text += "\n\n"

        return text

    def get_description_text(self):
        text = f"{self.subtitle}\n\n"
        text += self._get_workout_set_text("Easy")
        #text += self._get_workout_set_text("Medium")
        #text += self._get_workout_set_text("Hard")
        return text

    def __repr__(self):
        text = self.get_full_title() + "\n"
        text += self.get_description_text() + "\n"

        return text

    def append_workout_step(self, easy_workout, medium_workout, hard_workout):
        self.easy_set.append(easy_workout)
        self.medium_set.append(medium_workout)
        self.hard_set.append(hard_workout)


class PullUpChallenge:

    def __init__(self, week_no=1, start_date=date.today(), workout_frequecy=timedelta(days=2)):
        self._week_no = 1
        self.workout_start = start_date
        self.workout_frequency = workout_frequecy

    def _get_all_workouts(self):
        table_body = self._fetch_page_and_parse_table()

        rows = table_body.find_all('tr')
        all_workouts = []
        for row in rows:
            cols = row.find_all('td')
            if not cols:
                text = row.text
                text = text.splitlines()
                text = [line.strip() for line in text if line.strip()]
                assert len(text) == 2
                title = text[0]
                assert title.startswith("Day")
                subtitle = text[1]
                assert subtitle.startswith("Rest")
                workout_day = Workout(self._week_no, title, subtitle)
                all_workouts.append(workout_day)

            else:
                assert workout_day
                cols = [ele.text.strip() for ele in cols]
                assert len(cols) == 3
                workout_day.append_workout_step(*cols)

        return all_workouts

    def _fetch_page_and_parse_table(self):
        page = self._fetch_page()
        return self._parse_table(page)

    def _fetch_page(self):
        request = requests.get('https://twentypullups.com/')
        soup = BeautifulSoup(request.text, 'html.parser')
        return soup

    def _parse_table(self, html_soup):
        week_no_text = f"Week {self._week_no}"
        # link to week
        weeks = [i.get('href') for i in html_soup.find_all('a') if week_no_text in i.text]

        # parse table
        week_page_request = requests.get(weeks[0])
        week_page = BeautifulSoup(week_page_request.text, 'html.parser')

        table = week_page.find('table')
        table_body = table.find('tbody')

        return table_body

    def _workouts_to_cal_events(self):
        workouts = self._get_all_workouts()
        cal_events = []
        workout_date = self.workout_start
        for workout in workouts:
            title = workout.get_full_title()
            description = workout.get_description_text()
            event = CalendarEvent(title, workout_date, description)
            workout_date = workout_date + self.workout_frequency
            cal_events.append(event)
        return cal_events

    def _create_workouts_from_text(self):
        events = self._workouts_to_cal_events()
        cal = WorkoutCalendar(events)
        return cal.get_all_as_ical()

    def create_workout_calendar(self):
        cal_text = self._create_workouts_from_text()
        with open("pullup.ics", 'w') as out_file:
            out_file.writelines(cal_text)


if __name__ == "__main__":
    challenge = PullUpChallenge()
    challenge.create_workout_calendar()