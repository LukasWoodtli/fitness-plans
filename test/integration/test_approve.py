from datetime import date
from unittest.mock import patch
from approvaltests import Options, verify
from approvaltests.scrubbers import create_regex_scrubber

from fitness_plans.fitmacher_formel import FitMacherFormel
from fitness_plans.pullup_challenge import PullUpChallenge

OPTIONS_WITH_SCRUBBER = Options().with_scrubber(create_regex_scrubber("DTSTAMP;VALUE=DATE-TIME.*", "[creation date]"))


def test_fitmacher_formel():
    fmf = FitMacherFormel(date(2020, 5, 29))
    cal = fmf._create_workouts_from_text()
    verify(cal, options=OPTIONS_WITH_SCRUBBER)


def test_pullup_challenge():
    challenge = PullUpChallenge(start_date=date(2020, 5, 30))
    calendar = challenge._create_workouts_from_text()
    verify(calendar, options=OPTIONS_WITH_SCRUBBER)


@patch('fitness_plans.fitmacher_formel.FitMacherFormel._read_input_file')
def test_create_workout_calendar(mock_read_input_file):
    mock_read_input_file.return_value = \
        "1. TAG: Hello 1\n" \
        "Workout 1\n" \
        "2. TAG: Hello 2\n" \
        "Workout 2\n" \
        "3. TAG: Hello 3\n" \
        "Workout 3"
    fmf = FitMacherFormel()
    calendar = fmf._create_workouts_from_text()
    verify(calendar, options=OPTIONS_WITH_SCRUBBER)
