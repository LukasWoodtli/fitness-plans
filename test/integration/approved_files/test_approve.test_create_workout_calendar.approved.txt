BEGIN:VCALENDAR
VERSION:2.0
PRODID:Fitness
BEGIN:VEVENT
SUMMARY:1. TAG: Hello 1
DTSTART;VALUE=DATE:20211011
DTEND;VALUE=DATE:20211011
[creation date]
DESCRIPTION:1. TAG: Hello 1\nWorkout 1\n
END:VEVENT
BEGIN:VEVENT
SUMMARY:2. TAG: Hello 2
DTSTART;VALUE=DATE:20211012
DTEND;VALUE=DATE:20211012
[creation date]
DESCRIPTION:2. TAG: Hello 2\nWorkout 2\n
END:VEVENT
BEGIN:VEVENT
SUMMARY:3. TAG: Hello 3
DTSTART;VALUE=DATE:20211013
DTEND;VALUE=DATE:20211013
[creation date]
DESCRIPTION:3. TAG: Hello 3\nWorkout 3
END:VEVENT
END:VCALENDAR
